plumbing/format/diff/patch.go

```diff
@@ -0,0 +1,53 @@
+package diff
+
+import (
+	"io"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/filemode"
+)
+
+type Operation int
+
+const (
+	Equal Operation = iota
+	Add
+	Delete
+)
+
+// Patch represents the necessary steps to transform several files to another.
+type Patch interface {
+	FilePatches() []FilePatch
+	Message() string
+}
+
+// Patcheable is a Patch but with several ways to encode it. It is used to be
+// able to use Patch in a fluent way.
+type Patcheable interface {
+	Patch
+	// UnifiedDiff writes the patch into the provided writer in unified
+	// diff format. ctxLines changes the amount of unchanged lines that will be
+	// printed before and after each chunk.
+	UnifiedDiff(ctxLines int, w io.Writer) error
+}
+
+// FilePatch represents the necessary steps to transform one file to another
+type FilePatch interface {
+	IsBinary() bool
+	Files() (from, to File)
+	Chunks() []Chunk
+}
+
+// File contains all the file metadata necessary to print some diff formats
+type File interface {
+	Hash() plumbing.Hash
+	Mode() filemode.FileMode
+	Path() string
+}
+
+// Chunk is a operation to process to transform a file to another.
+// Operations can be Equal, Add or Delete
+type Chunk interface {
```

Comment:
Is called Type because is the Type of the Chunk. The type of the chunk is defined by an Operation.